### ISSUE TITLE
Can't call some python callables

### ISSUE THREAD

**Issue by t-kalinowski** [rstudio/reticulate#1024]
Hi,

Should R functions converted to python still be callable from R?

``` r
library(reticulate)
fn <- r_to_py(function(x) x)
try(fn(1)) # should this succeed?
#> Error in fn(1) : could not find function "fn"
```

Calling `py_to_r` enables calling again, but doesn’t change the class or printed output.

``` r
fn
#> <function make_python_function.<locals>.python_function at 0x7f243344aef0>
class(fn)
#> [1] "python.builtin.function" "python.builtin.object"
fn <- py_to_r(fn)
fn
#> <function make_python_function.<locals>.python_function at 0x7f243344aef0>
class(fn)
#> [1] "python.builtin.function" "python.builtin.object"
fn(1)
#> 1.0
```

(Changing the value of `convert` seems to make no difference.)

[This comment](https://github.com/rstudio/reticulate/blob/4967abddb35865a1be6ad839298695f1481bcade/R/python.R#L220) and the extensive use of `as.environment` suggests that maybe the intent is for all python callables to be callable from R and this is a bug?

It would be very convenient if all python callables are always directly callable from R without needing to go though `py_call`.

<sup>Created on 2021-07-02 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by kevinushey** [comment]
I don't think so, because Python objects (as represented in `reticulate`) are just R environments with a special class, and so cannot be plainly called.

Any callable Python function is really just a wrapper R function over the underlying Python object; e.g.

```
library(reticulate)

builtins <- import_builtins(convert = TRUE)
abs <- builtins$abs
typeof(abs)
unclass(abs)
```

and you should see something like:

```
> abs <- builtins$abs
> class(abs)
[1] "python.builtin.builtin_function_or_method"
[2] "python.builtin.object"                    
> typeof(abs)
[1] "closure"
> typeof(abs)
[1] "closure"
> unclass(abs)
function(...) {
    
    dots <- py_resolve_dots(list(...))
    result <- py_call_impl(callable, dots$args, dots$keywords)
    
    if (convert)
      result <- py_to_r(result)
      
    if (is.null(result))
      invisible(result)
    else
      result
    
  }
<bytecode: 0x62500cc631c8>
<environment: 0x62500ba6c620>
attr(,"py_object")
<built-in function abs>
```

I'm not sure if it would be possible for us to "skip" this R function wrapper in a way that Python objects could be directly callable.

**Comment by t-kalinowski** [comment]
Thanks. Would there be any downsides to indiscriminately wrapping all callable python objects in this R closure?

**Comment by kevinushey** [comment]
Off the top of my head, I don't think so, other than the implementation complexity (since now we'd need to handle unwrapping and re-wrapping of both Python-objects-as-environments and Python-objects-as-functions).
